### PR TITLE
Feature/updated dashboard

### PR DESCRIPTION
**Pull request by frenchbread**
Closes: #1040, #1105, #1066, #1145, #1144, #1245.

### CODE REVIEW

**Comment by frenchbread** [comment]
@apinf/developers Please review.

File: client/views/dashboard/charts/charts.js [hunk 1]
```diff
@@ -1,423 +1,336 @@
...
+
+  const instance = this;
+
+  instance.esData = new ReactiveVar(); // Handles ES data for charts
+  instance.tableDataSet = new ReactiveVar([]); // Handles parsed data for charts
+
+  instance.timeStart = new Date().getTime(); // Timer
+
+  const params = {
+    size: 50000,
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Should this still be hard-coded?

File: client/views/dashboard/charts/charts.js [hunk 2]
```diff
@@ -1,423 +1,336 @@
...
+    body: {
+      query: {
+        filtered: {
+          query: {
+            match_all: {}
+          },
+          filter: {
+            range: {
+              request_at: {
+                gte: moment().subtract(30, 'day').valueOf()
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: How can we parameterize this?

File: client/views/dashboard/charts/charts.js [hunk 3]
```diff
@@ -1,423 +1,336 @@
...
+      fields: [
+        'request_at',
+        'response_status',
+        'response_time',
+        'request_ip_country',
+        'request_ip',
+        'request_path',
+        'request_ip_location.lon',
+        'request_ip_location.lat',
+        'api_key'
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: We probably shouldnt include the `api_key` field in the analytics, since it is private. I.e. only the individual user should know about their API key.

Instead, use the `user_id` field, since it is unique.

File: client/views/dashboard/charts/charts.js [hunk 4]
```diff
@@ -1,423 +1,336 @@
...
     }
+  }
 
-  });
+  Meteor.call('getElasticSearchData', params, (err, res) => {
 
+    if (err) throw new Meteor.error(err);
 
-};
+    console.log('Got result!');
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Clean up console.log()s

File: client/views/dashboard/charts/charts.js [hunk 5]
```diff
@@ -1,423 +1,336 @@
...
 
-  // function that renders charts
   instance.renderCharts = function (parsedData) {
 
-    var timeStampDimension  = parsedData.timeStampDimension;
-    var timeStampGroup      = parsedData.timeStampGroup;
-    var timeScale           = parsedData.timeScale;
-    var overviewChart = dc.barChart("#overview-chart");
-    var moveChart = dc.barChart("#move-chart");
+    const {
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Nice destructuring!

File: client/views/dashboard/charts/charts.js [hunk 6]
```diff
@@ -1,423 +1,336 @@
...
 
-    // Parse each item adding timestamp as YMD and count the items
-    items.forEach(function (d) {
-      // Parse timestamp to Moment .jsobject
-      var timeStamp = moment(d.fields.request_at[0]);
+    const statusCodeGroup = statusCodeDimension.group(); // Create status code group
 
-      // Format the timestamp using Moment.js format method
-      timeStamp = timeStamp.format();
+    const binwidth = 100; // Init binwidth for a bar chart
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Consider changing this to 50, to see how it looks.

File: client/views/dashboard/charts/charts.js [hunk 7]
```diff
@@ -1,423 +1,336 @@
...
+      responseTimeDimension,
+      responseTimeGroup,
+      timeScaleForLine,
+      timeScaleForFocus,
+      xScaleForBar,
+      binwidth
+    } = parsedData;
+
+    // Init charts
+    const line = dc.lineChart('#line-chart');
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Can we give these more semantic names? E.g. `requestsOverTime`, `responseCodeCounts`, `responseTimeDistribution`

File: client/views/dashboard/charts/charts.html [hunk 8]
```diff
@@ -1,21 +1,35 @@
...
+        <div>
+          <h4 >
+            API Request Timeline
+            <small class="pull-right" style="line-height: 25px;">
+              <span id="row-selection" style="margin:0">
+                Showing <strong class="filter-count">_</strong> of <strong class="total-count">_</strong> records
+              </span>
+            </small>
+          </h4>
+          <div id="line-chart"></div>
```
Comment (brylie) [inline client/views/dashboard/charts/charts.html]: Change these IDs to match the chart names, if you update the chart names. See below.

File: client/views/dashboard/charts/charts.html [hunk 9]
```diff
@@ -1,21 +1,35 @@
...
+        <div>
+          <div id="focus-chart"></div>
+        </div>
+      </div>
     </div>
-    <div class="col-lg-12 col-md-12 col-sm-12">
-      {{> overviewChart}}
+    <div class="row">
+      <div class="col-md-6">
+        <h4>HTTP Status Codes</h4>
```
Comment (brylie) [inline client/views/dashboard/charts/charts.html]: Add i18n token.

File: client/views/dashboard/charts/charts.html [hunk 10]
```diff
@@ -1,21 +1,35 @@
...
+        <div id="row-chart"></div>
+      </div>
+      <div class="col-md-6">
+        <h4 >
+          API Response Time
```
Comment (brylie) [inline client/views/dashboard/charts/charts.html]: Add i18n token.

File: client/views/dashboard/charts/charts.js [hunk 11]
```diff
@@ -1,423 +1,330 @@
...
+Template.chartsLayout.onRendered(function () {
 
-    });
+  const instance = this;
 
-    // returns heat data
-    return addressPoints;
-  };
+  // Get reference to chart html elemets
+  const chartElemets = $('#line-chart, #focus-chart, #row-chart, #bar-chart');
```
Comment (brylie) [inline client/views/dashboard/charts/charts.js]: Remember to change this line if you change the chart names and template element IDs.

File: client/views/dashboard/charts/table/table.html [hunk 12]
```diff
@@ -1,31 +1,72 @@
 <template name="dataTable">
-  <div class="box box-primary" id="data-table-box">
-    <div class="box-header with-border">
-      <h3 class="box-title">{{_ "dataTable_Title"}}</h3>
-    </div>
-    <table class="datatable table table-bordered">
-      <thead>
+  <h4>Analytics data</h4>
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 13]
```diff
@@ -1,31 +1,72 @@
...
+  <table class="dataTable table table-bordered">
+    <thead>
+      <tr>
+        <th>
+          Date
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 14]
```diff
@@ -1,31 +1,72 @@
...
+        </th>
+        <th>
+          Country
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 15]
```diff
@@ -1,31 +1,72 @@
...
+        </th>
+        <th>
+          Path
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 16]
```diff
@@ -1,31 +1,72 @@
...
+        </th>
+        <th>
+          IP
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 17]
```diff
@@ -1,31 +1,72 @@
...
+        </th>
+        <th>
+          Response time (ms)
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 18]
```diff
@@ -1,31 +1,72 @@
...
-      </thead>
-      <tbody>
-
-      </tbody>
-    </table>
+      {{ / each}}
+    </tbody>
+  </table>
+  <div class="page-numbers pull-left">
+    <strong>{{ currentPageNumber }}</strong> of <strong>{{ totalPageNumber }}</strong>
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: We need to figure out how to internationalize this string. Lets bring it up during stand-up.

File: client/views/dashboard/charts/table/table.html [hunk 19]
```diff
@@ -1,31 +1,72 @@
...
+      </button>
+    {{ else }}
+      <button class="btn btn-sm btn-default disabled">
+        <span class="glyphicon glyphicon-chevron-left"></span>
+        Prev
+      </button>
+    {{ / if }}
+    {{ # if showNextButton }}
+      <button class="btn btn-sm btn-default" id="next">
+        Next
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 20]
```diff
@@ -1,31 +1,72 @@
...
+    </tbody>
+  </table>
+  <div class="page-numbers pull-left">
+    <strong>{{ currentPageNumber }}</strong> of <strong>{{ totalPageNumber }}</strong>
+  </div>
+  <div class="btn-group pull-right">
+    {{ # if showPrevButton }}
+      <button class="btn btn-sm btn-default" id="prev">
+        <span class="glyphicon glyphicon-chevron-left"></span>
+        Prev
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 21]
```diff
@@ -1,31 +1,72 @@
...
+      </button>
+    {{ else }}
+      <button class="btn btn-sm btn-default disabled">
+        <span class="glyphicon glyphicon-chevron-left"></span>
+        Prev
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.html [hunk 22]
```diff
@@ -1,31 +1,72 @@
...
+      </button>
+    {{ / if }}
+    {{ # if showNextButton }}
+      <button class="btn btn-sm btn-default" id="next">
+        Next
+        <span class="glyphicon glyphicon-chevron-right"></span>
+      </button>
+    {{ else }}
+      <button class="btn btn-sm btn-default disabled">
+        Next
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.html]: Add i18n token.

File: client/views/dashboard/charts/table/table.js [hunk 23]
```diff
@@ -0,0 +1,73 @@
+Template.dataTable.onCreated(function () {
+
+  const instance = this;
+
+  instance.rowCount = new ReactiveVar(10);
+  instance.pageNumber = new ReactiveVar(1);
+
+  // instance.autorun(() => {
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Cleanup.

File: client/views/dashboard/charts/table/table.js [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+
+});
+
+Template.dataTable.events({
+  'click #prev': function (event, instance) {
+
+    let oldPageNumber = instance.pageNumber.get();
+
+    if (oldPageNumber > 1) {
+      instance.pageNumber.set(oldPageNumber-1);
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Use whitespace around subtraction operator.

File: client/views/dashboard/charts/table/table.js [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  },
+  'click #next': function (event, instance) {
+
+    let oldPageNumber = instance.pageNumber.get()
+
+    instance.pageNumber.set(oldPageNumber+1);
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Add whitespace around addition operator.

File: client/views/dashboard/charts/table/table.js [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+  }
+})
+
+Template.dataTable.helpers({
+  tableDataSet () {
+
+    const instance = Template.instance();
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Add comments here, even though these variable names are clear.

File: client/views/dashboard/charts/table/table.js [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+  showPrevButton () {
+
+    const instance = Template.instance();
+
+    const pageNumber = instance.pageNumber.get();
+
+    return pageNumber > 1;
+  },
+  showNextButton () {
+
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Add clarifying comments.

File: client/views/dashboard/charts/charts.js (hunk 7 above)
Comment (frenchbread) [inline client/views/dashboard/charts/charts.js]: In scope of this file and charts these names technically related to the type of a chart but not what it represents.

File: client/views/dashboard/charts/table/table.js [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+    const rowCount = instance.rowCount.get();
+    const pageNumber = instance.pageNumber.get();
+
+    const arrStart = rowCount * pageNumber;
+    const arrEnd = arrStart + rowCount;
+
+    return Template.currentData().tableDataSet.slice(arrStart, arrEnd);
+  },
+  showPrevButton () {
+
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Add descriptive comments, higher level describing the purpose of this function.

File: client/views/dashboard/charts/table/table.js [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+    const instance = Template.instance();
+
+    const rowCount = instance.rowCount.get();
+    const pageNumber = instance.pageNumber.get();
+
+    const dataSetLength = Template.currentData().tableDataSet.length;
+
+    return pageNumber < (dataSetLength / rowCount - 1);
+  },
+  currentPageNumber () {
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Describe this function with a comment.

File: client/views/dashboard/charts/table/table.js [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+    const instance = Template.instance();
+
+    return instance.pageNumber.get();
+  },
+  totalPageNumber () {
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Describe this function with a comment.

File: client/views/dashboard/charts/table/table.js [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+    const instance = Template.instance();
+
+    return (Template.currentData().tableDataSet.length / instance.rowCount.get()) | 0;
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.js]: Describe what is going on here. This is really long and should probably be broken into a couple of steps with meaningful variable names.

File: client/views/dashboard/charts/table/table.less [hunk 32]
```diff
@@ -1,8 +1,18 @@
...
 
-.box-header.with-border {
-  margin-bottom: 1em;
+table.dataTable.table-bordered, table.dataTable.table-bordered > tbody > tr > td {
+  border: 1px solid #eee;
+}
+
+
+.padding-sides {
+  padding: 0 15px;
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.less]: Use relative values for layout rules.

File: client/views/dashboard/charts/table/table.less [hunk 33]
```diff
@@ -1,8 +1,18 @@
...
+}
+
+.page-numbers {
+  line-height: 30px;
```
Comment (brylie) [inline client/views/dashboard/charts/table/table.less]: Use relative values (ems or rems) for layout style rules.

File: client/views/dashboard/charts/charts.js (hunk 1 above)
Comment (frenchbread) [inline client/views/dashboard/charts/charts.js]: @brylie I think so for now. Otherwise we could add an input field to the UI, to allow user provide custom limit. That will probably affect design of the wireframes.

File: client/views/dashboard/charts/charts.js (hunk 2 above)
Comment (frenchbread) [inline client/views/dashboard/charts/charts.js]: @brylie Yep, I suggest to do it in separate PR.

File: client/views/dashboard/charts/table/table.html (hunk 18 above)
Comment (frenchbread) [inline client/views/dashboard/charts/table/table.html]: We could internationalize  "of" word for now. But also it's possible to add page numbers as a button group which will exclude use of any words here. E.g. `[1][2][3]...[10]`.

Comment (bajiat) [inline client/views/dashboard/charts/table/table.html]: I strongly advice against having 1 of 20 and internationalising only of. Not all languages have prepositions and Finnish is one of those languages: we add endings to words. One of the key things in internationalisation is never to include only sentence fragments as localisable entities, Either internationalise the whole thing, in which case page numbers are parameters that get automatically added or find another solution. Button group is one option. Another option would be to use / instead of the word 'of'

File: client/views/dashboard/charts/bar/bar.html [hunk 34]
```diff
@@ -1,11 +1,4 @@
...
-    <div class="box-header with-border">
-      <h3 class="box-title">{{_ "overviewChart_Title"}}</h3>
-    </div>
-    <div class="box-body">
-      <div id="overview-chart">
-      </div>
-    </div>
-  </div>
+<template name="statusCodeCountsChart">
+  <h4>HTTP Status Codes</h4>
```
Comment (brylie) [inline client/views/dashboard/charts/bar/bar.html]: Add i18n token here.

File: client/views/dashboard/charts/charts.html [hunk 35]
```diff
@@ -1,35 +1,28 @@
...
       </div>
     </div>
+  </div>
+  <div class="row">
+    <div class="col-md-6">
+      {{ > statusCodeCountsChart }}
+    </div>
+    <div class="col-md-6">
+      <h4 >
+        API Response Time
```
Comment (brylie) [inline client/views/dashboard/charts/charts.html]: Add i18n token here.

File: client/views/dashboard/charts/line/line.html [hunk 36]
```diff
@@ -0,0 +1,11 @@
+<template name="requestsOverTimeChart">
+  <h4 >
+    API Request Timeline
+    <small class="pull-right" style="line-height: 25px;">
+      <span id="row-selection" style="margin:0">
+        Showing <strong class="filter-count">_</strong> of <strong class="total-count">_</strong> records
```
Comment (brylie) [inline client/views/dashboard/charts/line/line.html]: Lets remember to add i18n token here. This is a tricky string, since it is broken into multiple pieces. E.g. different languages have different word order, so the variables (filter count and total count) would be in different parts of the phrase. Try to simplify this string, so that the internationalized text is separate from the variable text.

**Comment by frenchbread** [comment]
Added suggested changes.

**Comment by brylie** [comment]
This looks good. @nnn would you like to review this PR, since it is related to your re-design work?

**Comment by nnn** [comment]
I have merged develop. Looks good. All other improvements I will add as a separate PR.

**Comment by nnn** [comment]
Merging this. There is going to a separate task to finalize this admin dashboard look.
@frenchbread, great work!
